warehouse/macaroons/models.py

```diff
@@ -37,13 +38,27 @@ class Macaroon(db.Model):
         UniqueConstraint(
             "description", "user_id", name="_user_macaroons_description_uc"
         ),
+        UniqueConstraint(
+            "description", "project_id", name="_project_macaroons_description_uc"
+        ),
+        CheckConstraint(
+            "(user_id::text IS NULL) <> (project_id::text IS NULL)",
+            name="_user_xor_project_macaroon",
+        ),
     )
 
-    # All of our Macaroons belong to a specific user, because a caveat-less
-    # Macaroon should act the same as their password does, instead of as a
-    # global permission to upload files.
+    # Macaroons come in two forms: they either belong to a user, or they
+    # belong to a project.
+    # * In the user case, a Macaroon has an associated user, and *might* have
+    #   additional project scope restrictions as part of its caveats.
+    # * In the project case, a Macaroon has an associated project, and
+    #   is scoped to just that project.
     user_id = Column(
-        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
+        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, index=True
+    )
+
+    project_id = Column(
+        UUID(as_uuid=True), ForeignKey("projects.id"), nullable=True, index=True
```

Comment:
Flagging: I'm not sure if `index=True` is necessary here. I just added it because we have it for the `user_id` FK.